ariadne/src/Ariadne/Wallet/Backend/KeyStorage.hs

```diff
@@ -132,114 +144,126 @@ runCatchInState m = StateT $ \s ->
     Left e -> (Left e, s)
     Right (a, s') -> (Right a, s')
 
--- | Get the wallet index by name or using current selection.
+-- | Get the wallet HdRootId by name or using current selection.
+-- resolveWalletRef
 resolveWalletRef
   :: IORef (Maybe WalletSelection)
-  -> (CardanoMode ~> IO)
   -> WalletReference
-  -> IO Word
-resolveWalletRef walletSelRef runCardanoMode = \case
+  -> DB
+  -> IO HdRootId
+resolveWalletRef walRef walletSelRef db = case walRef of
   WalletRefSelection -> do
     mWalletSelection <- readIORef walletSelRef
     case mWalletSelection of
       Nothing -> throwM NoWalletSelection
-      Just WalletSelection{..} -> return wsWalletIndex
-  WalletRefByName (WalletName name) -> do
-    us <- runCardanoMode getSecretDefault
-    case findIndex (\w -> w ^. wdName == name) (us ^. usWallets) of
-      Just i -> return (fromIntegral i)
-      Nothing -> throwM $ WalletDoesNotExist name
-  WalletRefByIndex i -> return i
-
--- | Like 'resolveWalletRef', but for accounts.
-resolveAccountRef
-  :: IORef (Maybe WalletSelection)
-  -> (CardanoMode ~> IO)
-  -> AccountReference
-  -> IO (Word, Word32)
-resolveAccountRef walletSelRef runCardanoMode accountRef = do
-    walletIdx <- resolveWalletRef walletSelRef runCardanoMode walletRef
-    (walletIdx, ) <$> resolveAccountIdx walletIdx
+      Just WalletSelection{..} -> return $ getHdRootId wsPath
+  WalletRefByName walletName -> do
+    case getHdRootIdByName walletName hdWallets of
+      Just hdR -> return hdR
+      Nothing -> throwM $ WalletDoesNotExist walletName
+  WalletRefByIndex i -> do
+    -- Note: Add/remove wallets cause changes in indexation
```

Comment:
For now we agreed to have an assumption that the order of items in `IxSet`s is the same as the order in UI. We already had this assumption, but it was less noticeable, because backend was storing all data in lists/vectors.
In the future (I suppose after AD-91) I want to make frontend and backend use only stable identifiers. For example, `WalletRefByIndex i` will be processed by frontend which should know which data it displays and know how to convert this `i` to wallet identifier (which is hash of it's public key). Frontend then will call backend and pass stable identifier there.